src/ioctls/vm.rs

```diff
@@ -0,0 +1,416 @@
+
+/// Used in `VmFd::register_ioevent` to indicate that no datamatch is requested.
+pub struct NoDatamatch;
+impl Into<u64> for NoDatamatch {
+    fn into(self) -> u64 {
+        0
+    }
+}
+
+/// A wrapper around creating and using a VM.
+pub struct VmFd {
+    vm: File,
+    run_size: usize,
+}
+
+impl VmFd {
+    /// Creates/modifies a guest physical memory slot.
+    ///
+    /// See the documentation for `KVM_SET_USER_MEMORY_REGION`.
+    ///
+    pub fn set_user_memory_region(
+        &self,
+        user_memory_region: kvm_userspace_memory_region,
+    ) -> Result<()> {
+        let ret =
+            unsafe { ioctl_with_ref(self, KVM_SET_USER_MEMORY_REGION(), &user_memory_region) };
+        if ret == 0 {
+            Ok(())
+        } else {
+            Err(io::Error::last_os_error())
+        }
+    }
+
+    /// Sets the address of the three-page region in the VM's address space.
+    ///
+    /// See the documentation on the `KVM_SET_TSS_ADDR` ioctl.
+    ///
+    /// # Arguments
+    ///
+    /// * `offset` - Physical address of a three-page region in the guest's physical address space.
+    ///
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    pub fn set_tss_address(&self, offset: usize) -> Result<()> {
+        // Safe because we know that our file is a VM fd and we verify the return result.
+        let ret = unsafe { ioctl_with_val(self, KVM_SET_TSS_ADDR(), offset as c_ulong) };
+        if ret == 0 {
+            Ok(())
+        } else {
+            Err(io::Error::last_os_error())
+        }
+    }
+
+    /// Creates an in-kernel interrupt controller.
+    ///
+    /// See the documentation for `KVM_CREATE_IRQCHIP`.
+    ///
+    #[cfg(any(
+        target_arch = "x86",
+        target_arch = "x86_64",
+        target_arch = "arm",
+        target_arch = "aarch64"
+    ))]
+    pub fn create_irq_chip(&self) -> Result<()> {
+        // Safe because we know that our file is a VM fd and we verify the return result.
+        let ret = unsafe { ioctl(self, KVM_CREATE_IRQCHIP()) };
+        if ret == 0 {
+            Ok(())
+        } else {
+            Err(io::Error::last_os_error())
+        }
+    }
+
+    /// Creates a PIT as per the `KVM_CREATE_PIT2` ioctl.
+    ///
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    pub fn create_pit2(&self, pit_config: kvm_pit_config) -> Result<()> {
+        // Safe because we know that our file is a VM fd, we know the kernel will only read the
+        // correct amount of memory from our pointer, and we verify the return result.
+        let ret = unsafe { ioctl_with_ref(self, KVM_CREATE_PIT2(), &pit_config) };
+        if ret == 0 {
+            Ok(())
+        } else {
+            Err(io::Error::last_os_error())
+        }
+    }
+
+    /// Registers an event to be signaled whenever a certain address is written to.
+    ///
+    /// # Arguments
+    ///
+    /// * `evt` - EventFd which will be signaled. When signaling, the usual `vmexit` to userspace
```

Comment:
It's not an `EventFd` anymore. I think this was the last remaining `sys_util` dependency, thanks for removing it.

Reply:
True. Yap, I removed the dependencies on sys_util & memory_model. But we still need the ioctl. I will update the docs.

Reply:
Done.